File: Tree/1038. Binary Search Tree to Greater Sum Tree/bst_to_gst.go
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func bstToGst(root *TreeNode) *TreeNode {
	var sum int
	var traverse func(*TreeNode)
	traverse = func(node *TreeNode) {
		if node != nil {
			traverse(node.Right)
			sum += node.Val
			node.Val = sum
			traverse(node.Left)
		}
	}
	traverse(root)
	return root
}

File: Tree/1038. Binary Search Tree to Greater Sum Tree/bst_to_gst.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn bst_to_gst(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut sum = 0;
        Self::traverse(&root, &mut sum);
        root
    }

    fn traverse(node: &Option<Rc<RefCell<TreeNode>>>, sum: &mut i32) {
        if let Some(n) = node {
            let mut n = n.borrow_mut();
            Self::traverse(&n.right, sum);
            *sum += n.val;
            n.val = *sum;
            Self::traverse(&n.left, sum);
        }
    }
}